ldb/tx.go

```diff
@@ -346,3 +375,203 @@ func (db *LevelDb) FetchTxBySha(txsha *btcwire.ShaHash) ([]*btcdb.TxListReply, e
+		rawIndex := make([]byte, 22, 22)
+		copy(rawIndex, iter.Key()[22:])
+		addrIndex := unpackTxIndex(rawIndex)
+
+		tx, blkSha, blkHeight, _, err := db.fetchTxDataByLoc(addrIndex.blkHeight,
+			addrIndex.txoffset, addrIndex.txlen, []byte{})
+		if err != nil {
+			// Eat a possible error due to a potential re-org.
+			continue
+		}
+
+		txSha, _ := tx.TxSha()
+		txReply := &btcdb.TxListReply{Sha: &txSha, Tx: tx,
+			BlkSha: blkSha, Height: blkHeight, TxSpent: []bool{}, Err: err}
+
+		replies = append(replies, txReply)
+		limit--
+	}
+	iter.Release()
+
+	return replies, nil
+}
+
+// UpdateAddrIndexForBlock updates the stored addrindex with passed
+// index information for a particular block height. Additionally, it
+// will update the stored meta-data related to the curent tip of the
+// addr index. These two operations are performed in an atomic
+// transaction which is commited before the function returns.
+// Addresses are indexed by the raw bytes of their base58 decoded
+// hash160.
+// NOTE: See comment for addrIndexToKey for information pertaining to
+// how the address index is currently represented and stored within the
+// db.
+func (db *LevelDb) UpdateAddrIndexForBlock(blkSha *btcwire.ShaHash, blkHeight int64, addrIndex *btcdb.BlockAddrIndex) error {
+	db.dbLock.Lock()
+	defer db.dbLock.Unlock()
+
+	var blankData []byte
+	batch := db.lBatch()
+	defer db.lbatch.Reset()
+
+	// Write all data for the new address indexes in a single batch
+	// transaction.
+	for addrKey, indexes := range *addrIndex {
+		for _, txLoc := range indexes {
+			index := &txAddrIndex{
+				hash160:   addrKey,
+				blkHeight: blkHeight,
+				txoffset:  txLoc.TxStart,
+				txlen:     txLoc.TxLen,
+			}
+			// The index is stored purely in the key.
+			packedIndex := addrIndexToKey(index)
+			batch.Put(packedIndex, blankData)
+		}
+	}
+
+	// Update tip of addrindex.
+	newIndexTip := make([]byte, 40, 40)
+	copy(newIndexTip[:32], blkSha.Bytes())
+	binary.LittleEndian.PutUint64(newIndexTip[32:], uint64(blkHeight))
+	batch.Put(addrIndexMetaDataKey, newIndexTip)
+
+	if err := db.lDb.Write(batch, db.wo); err != nil {
+		return err
+	}
+
+	db.lastAddrIndexBlkIdx = blkHeight
+	db.lastAddrIndexBlkSha = *blkSha
+
+	return nil
+}
+
+// DeleteAddrIndex deletes the entire addrindex stored within the DB.
+// It also resets the cached in-memory metadata about the addr index.
+func (db *LevelDb) DeleteAddrIndex() error {
+	db.dbLock.Lock()
+	defer db.dbLock.Unlock()
+
+	batch := db.lBatch()
+	defer batch.Reset()
+
+	// Delete the entire index along with any metadata about it.
+	iter := db.lDb.NewIterator(bytesPrefix(addrIndexKeyPrefix), db.ro)
+	numInBatch := 0
+	for iter.Next() {
+		key := iter.Key()
+		batch.Delete(key)
+
+		numInBatch++
+
+		// Delete in chunks to potentially avoid very large batches.
+		if numInBatch >= batchDeleteThreshold {
+			if err := db.lDb.Write(batch, db.wo); err != nil {
+				return err
+			} else {
+				batch.Reset()
+				numInBatch = 0
+			}
```

Comment:
I agree, but I suspect lint didn't catch it because I imagine it only look before returns instead of in nested ifs like this.